### DISCUSSION TITLE
VCPKG_BINARY_SOURCES environment variable ignored when set in CMakeLists.txt

### DISCUSSION THREAD

**Discussion by omartijn** [microsoft/vcpkg#19329] [Q&A]
**Describe the bug**
Setting the `VCPKG_BINARY_SOURCES` environment variable is ignored when it is done from `CMakeLists.txt` to be used in manifest mode. It works when explicitly exported.

**Environment**
- OS: Tested on Linux
- Compiler: Irrelevant

**To Reproduce**
Create a CMakeLists.txt, add vcpkg as subdirectory, set the environment variable and then set the toolchain file

**Expected behavior**
vcpkg should honor the variable (i.e. use nuget when set)

**Failure logs** 
vcpkg will ignore the variable

**Comment by autoantwort** [comment]
As a workaround you can set [VCPKG_INSTALL_OPTIONS](https://github.com/microsoft/vcpkg/blob/master/docs/users/manifests.md#vcpkg_install_options)

**Comment by JackBoosY** [comment]
We recommend passing all the parameters related to vcpkg in the cmake command line, because the vcpkg tool chain will be called before CMakeLists.

**Comment by omartijn** [comment]
@autoantwort Thanks that indeed works around the issue.

@JackBoosY That's inconvenient. One of the main advantages of manifest mode - at least to me - is that it allows people to "just build the project", without even being aware that it uses vcpkg. If we then suddenly require the user to give extra flags on the cmake configure step, that kind of defeats the purpose.

**Comment by dg0yt** [comment]
> One of the main advantages of manifest mode - at least to me - is that it allows people to "just build the project", without even being aware that it uses vcpkg.

@omartijn AFAIK the user has to explicitly activate vcpkg integration before the manifest has any affect. And while I'm fine with accepting the dependencies from the manifest when vcpkg integration is active (*my* vcpkg root), I'm not sure if this should imply using the project's binary artifacts instead of those from [my environment](https://github.com/microsoft/vcpkg/blob/master/docs/users/config-environment.md).

**Comment by omartijn** [comment]
@dg0yt You can enable integration by adding something like this to your `CMakeLists.txt`:

`set(CMAKE_TOOLCHAIN_FILE ${CMAKE_CURRENT_SOURCE_DIR}/vcpkg/scripts/buildsystems/vcpkg.cmake CACHE STRING "Vcpkg toolchain file")`

That's assuming you have vcpkg checked out as a submodule in the `vcpkg` directory. Since some of our dependencies take quite long to build, however (and we also have our own private registries as well) it'd be nice if we could cache these binaries easily. This goes especially well with our CI setup, as those machines have more horsepower than our developer machines so developers can then use binary artefacts generated by the CI machines.

**Comment by dg0yt** [comment]
@omartijn It may work for you, but in general I would not recommend setting the `CMAKE_TOOLCHAIN_FILE` inside the project. It is intrusive, mixing independent concerns. It prompts for even more cmake code to offer to a way to opt out.
The canonical usage of cmake is to pass the toolchain as a parameter on first call of cmake. And probably you will have to set additional parameters or environment variables anyway.

**Comment by Hoikas** [comment]
> the vcpkg tool chain will be called before CMakeLists.

This is incorrect. The toolchain is called during the `project()` call in CMakeLists.

> The canonical usage of cmake is to pass the toolchain as a parameter on first call of cmake. And probably you will have to set additional parameters or environment variables anyway.

The README [suggests otherwise.](https://github.com/microsoft/vcpkg#vcpkg-as-a-submodule)

**Reply by Be-ing** [reply]
Indeed, setting VCPKG_BINARY_SOURCES before the `project()` call in CMakeLists.txt works fine. We do this in Tenacity: https://github.com/tenacityteam/tenacity/blob/master/CMakeLists.txt#L96

**Reply by jsanchez19** [reply]
ah, that this variable needs to be an environment variable is something I missed! I was setting it as a normal CMake variable and that wasn't working for me: In the link you provided, you're setting it as an environment variable within CMake. 

```cmake
set( ENV{VCPKG_BINARY_SOURCES} "clear")
```

**Comment by JackBoosY** [comment]
Or you can set these vcpkg related vars before `project()`.

**Comment by jsanchez19** [comment]
In my case setting the environment variable works, but setting it on the command line (`cmake -DVCPKG_BINARY_SOURCES=clear ...`) does not work. Setting it inside the CMakeLists.txt file, before `project()` also doesn't work.

I'm using manifest mode with a custom registry, and vcpkg version `2021-05-05-9f849c4c43e50d1b16186ae76681c27b0c1be9d9`
